pkg/objstore/tracing.go

```diff
@@ -0,0 +1,143 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package objstore
+
+import (
+	"context"
+	"io"
+
+	"github.com/opentracing/opentracing-go"
+
+	"github.com/thanos-io/thanos/pkg/tracing"
+)
+
+// TracingBucket includes bucket operations in the traces.
+type TracingBucket struct {
+	bkt Bucket
+}
+
+func NewTracingBucket(bkt Bucket) InstrumentedBucket {
+	return TracingBucket{bkt: bkt}
+}
+
+func (t TracingBucket) Iter(ctx context.Context, dir string, f func(string) error) (err error) {
+	tracing.DoWithSpan(ctx, "bucket_iter", func(spanCtx context.Context, span opentracing.Span) {
+		span.LogKV("dir", dir)
+		err = t.bkt.Iter(spanCtx, dir, f)
+	})
+	return
+}
+
+func (t TracingBucket) Get(ctx context.Context, name string) (io.ReadCloser, error) {
+	span, spanCtx := tracing.StartSpan(ctx, "bucket_get")
+	span.LogKV("name", name)
+
+	r, err := t.bkt.Get(spanCtx, name)
+	if err != nil {
+		span.LogKV("err", err)
+		span.Finish()
+		return nil, err
+	}
+
+	return &tracingReadCloser{r: r, s: span}, nil
+}
+
+func (t TracingBucket) GetRange(ctx context.Context, name string, off, length int64) (io.ReadCloser, error) {
+	span, spanCtx := tracing.StartSpan(ctx, "bucket_getrange")
+	span.LogKV("name", name, "offset", off, "length", length)
+
+	r, err := t.bkt.GetRange(spanCtx, name, off, length)
+	if err != nil {
+		span.LogKV("err", err)
+		span.Finish()
+		return nil, err
+	}
+
+	return &tracingReadCloser{r: r, s: span}, nil
+}
+
+func (t TracingBucket) Exists(ctx context.Context, name string) (exists bool, err error) {
+	tracing.DoWithSpan(ctx, "bucket_exists", func(spanCtx context.Context, span opentracing.Span) {
+		span.LogKV("name", name)
+		exists, err = t.bkt.Exists(spanCtx, name)
+	})
+	return
+}
+
+func (t TracingBucket) Attributes(ctx context.Context, name string) (attrs ObjectAttributes, err error) {
+	tracing.DoWithSpan(ctx, "bucket_attributes", func(spanCtx context.Context, span opentracing.Span) {
+		span.LogKV("name", name)
+		attrs, err = t.bkt.Attributes(spanCtx, name)
+	})
+	return
+}
+
+func (t TracingBucket) Upload(ctx context.Context, name string, r io.Reader) (err error) {
+	tracing.DoWithSpan(ctx, "bucket_upload", func(spanCtx context.Context, span opentracing.Span) {
+		span.LogKV("name", name)
+		err = t.bkt.Upload(spanCtx, name, r)
+	})
+	return
+}
+
+func (t TracingBucket) Delete(ctx context.Context, name string) (err error) {
+	tracing.DoWithSpan(ctx, "bucket_delete", func(spanCtx context.Context, span opentracing.Span) {
+		span.LogKV("name", name)
+		err = t.bkt.Delete(spanCtx, name)
+	})
+	return
+}
+
+func (t TracingBucket) Name() string {
+	return "tracing: " + t.bkt.Name()
```

Comment:
Not sure if we should rename this. Why not just returning `t.bkt.Name()`? We're just adding instrumentation, without any logic change.

Reply:
This method is only called by metrics bucket, which is not used on top of tracking bucket. Personally I would suggest removing this method completely.